Add masonry as valid value to grid-template-rows/grid-template-columns.

sammygill opened the pull request:
#### ec9286482794e9be0e62138642fefc3b67d378be
<pre>
Add masonry as valid value to grid-template-rows/grid-template-columns.
<a href="https://bugs.webkit.org/show_bug.cgi?id=246541">https://bugs.webkit.org/show_bug.cgi?id=246541</a>
rdar://101188057

Reviewed by Antti Koivisto.

This patch adds a CSS Masonry Layout flag and adds the &apos;masonry&apos;
value as a valid value for the grid-template-rows and
grid-template-columns properties. A new CSSMasonryValue has been added
to represent this new value.

When we parse the grid-template-rows/columns property, we check to
see if the token is of CSSValueMasonry and consume it as an ident if it
is since the value of the syntax is straightforward.

When we build the track list, we check the ID of the primitive value
to see if it also matches CSSMasonryValue. In this scenario a new
masonry entry is added to the track list so that the serialization can
later set m_masonryRows or m_masonryColumns in StyleGridData correctly.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/masonry-grid-template-columns-computed-withcontent-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/masonry-parsing-expected.txt:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::canInterpolate):
(WebCore::blendFunc):
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForGridTrackListi):
If the track list is empty and the value for the track is not subgrid,
then we check to see if it is set to masonry.

* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::operator==):
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeGridTemplatesRowsOrColumns):
Since &apos;masonry&apos; is the only value that will trigger masonry layout and
there is no other form for the syntax, we can just consume the token
as an ident. However, later on when we actually build the track list,
we will need to check whether the value is a CSSPrimitive and compare
the ID of it.

* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::gridMasonryRows const):
(WebCore::RenderStyle::gridMasonryColumns const):
* Source/WebCore/rendering/style/StyleGridData.cpp:
(WebCore::StyleGridData::StyleGridData):
(WebCore::StyleGridData::setRows):
(WebCore::StyleGridData::setColumns):
(WebCore::StyleGridData::computeCachedTrackData):
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/StyleGridData.h:
(WebCore::GridTrackEntryMasonry::operator== const):
(WebCore::StyleGridData::masonryRows const):
(WebCore::StyleGridData::masonryColumns const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::createGridTrackList):

Canonical link: <a href="https://commits.webkit.org/255888@main">https://commits.webkit.org/255888@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5216a53ac588578d08880bc7026566e45f211d2c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93756 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2951 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24356 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103400 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163721 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97749 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2967 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31205 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86093 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99427 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99415 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2113 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80195 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29142 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84039 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/83763 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72097 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/85012 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37599 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17593 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/80139 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35449 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18853 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27657 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4065 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39327 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41412 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/82782 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41260 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38084 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18715 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebCore/css/CSSMasonryValue.h:
```diff
@@ -0,0 +1,46 @@
+
+namespace WebCore {
+
+class CSSMasonryValue final : public CSSValue {
```
Do we need this new class? It doesn't appear to be used anywhere.

The parser is generating a CSSPrimitiveValue, with the CSSValueMasonry enum value inside it.

sammygill replied:
Removed

mattwoodrow commented on Source/WebCore/rendering/style/StyleGridData.h:
```diff
@@ -168,6 +177,9 @@ class StyleGridData : public RefCounted<StyleGridData> {
     bool m_subgridRows;
     bool m_subgridColumns;
 
+    bool m_masonryRows;
```
Might be worth combining this into m_subgridRows with a tri-state enum (normal, subgrid, masonry), so that the invalid state of both masonry+subgrid can't be represented.

mdubet commented on Source/WebCore/css/CSSMasonryValue.h:
```diff
@@ -0,0 +1,46 @@
+
+namespace WebCore {
+
+class CSSMasonryValue final : public CSSValue {
```
What is the goal of this new CSSValue class instead of just keeping `masonry` as a CSSValueID ?

sammygill replied:
We actually don't need it. It was mostly because it was not clear to me what was being generated/used, but @mattwoodrow helped clear it up for me so I'll be removing it.

sammygill replied:
Removed

anttijk commented on Source/WebCore/rendering/style/StyleGridData.h:
```diff
@@ -117,6 +123,9 @@ class StyleGridData : public RefCounted<StyleGridData> {
     const bool& subgridColumns() const { return m_subgridColumns; }
 
+    const bool& masonryRows() const { return m_masonryRows; }
+    const bool& masonryColumns() const { return m_masonryColumns; }
```
Do these variables need to be added to operator==?

Why do these function (and some earlier ones) return references instead of just bools?

sammygill replied:
Yes they probably should be. Added!

These did not need to so I went ahead and changed the methods, but I'm not familiar with the ones above.

sammygill committed:
Add masonry as valid value to grid-template-rows/grid-template-columns.
https://bugs.webkit.org/show_bug.cgi?id=246541
rdar://101188057

Reviewed by Antti Koivisto.

This patch adds a CSS Masonry Layout flag and adds the 'masonry'
value as a valid value for the grid-template-rows and
grid-template-columns properties. A new CSSMasonryValue has been added
to represent this new value.

When we parse the grid-template-rows/columns property, we check to
see if the token is of CSSValueMasonry and consume it as an ident if it
is since the value of the syntax is straightforward.

When we build the track list, we check the ID of the primitive value
to see if it also matches CSSMasonryValue. In this scenario a new
masonry entry is added to the track list so that the serialization can
later set m_masonryRows or m_masonryColumns in StyleGridData correctly.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/masonry-grid-template-columns-computed-withcontent-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/masonry-parsing-expected.txt:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::canInterpolate):
(WebCore::blendFunc):
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForGridTrackListi):
If the track list is empty and the value for the track is not subgrid,
then we check to see if it is set to masonry.

* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::operator==):
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeGridTemplatesRowsOrColumns):
Since 'masonry' is the only value that will trigger masonry layout and
there is no other form for the syntax, we can just consume the token
as an ident. However, later on when we actually build the track list,
we will need to check whether the value is a CSSPrimitive and compare
the ID of it.

* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::gridMasonryRows const):
(WebCore::RenderStyle::gridMasonryColumns const):
* Source/WebCore/rendering/style/StyleGridData.cpp:
(WebCore::StyleGridData::StyleGridData):
(WebCore::StyleGridData::setRows):
(WebCore::StyleGridData::setColumns):
(WebCore::StyleGridData::computeCachedTrackData):
(WebCore::operator<<):
* Source/WebCore/rendering/style/StyleGridData.h:
(WebCore::GridTrackEntryMasonry::operator== const):
(WebCore::StyleGridData::masonryRows const):
(WebCore::StyleGridData::masonryColumns const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::createGridTrackList):

Canonical link: https://commits.webkit.org/255888@main
